File: results/dnd-rest-benchmark/lifecycle-runs/20260731T052954Z_claude_claude-sonnet-5_python-flask/dndapp/routes/audit.py
"""Deterministic audit log and export summaries for campaign state."""

from flask import Blueprint, jsonify

from ..db import SCHEMA_VERSION, get_db

bp = Blueprint("audit", __name__)


def _count(conn, table, campaign_id):
    return conn.execute(
        f"SELECT COUNT(*) AS c FROM {table} WHERE campaign_id = ?",
        (campaign_id,),
    ).fetchone()["c"]


@bp.get("/v1/campaigns/<campaign_id>/audit")
def get_campaign_audit(campaign_id):
    conn = get_db()
    try:
        campaign = conn.execute(
            "SELECT id FROM campaigns WHERE id = ?", (campaign_id,)
        ).fetchone()
        if campaign is None:
            return jsonify(error="campaign not found"), 404

        events = _count(conn, "campaign_events", campaign_id)
        quests = _count(conn, "quests", campaign_id)
        npcs = _count(conn, "npcs", campaign_id)
        sessions = _count(conn, "campaign_sessions", campaign_id)
    finally:
        conn.close()

    return jsonify(
        campaign_id=campaign_id,
        events=events,
        quests=quests,
        npcs=npcs,
        sessions=sessions,
    )


@bp.get("/v1/campaigns/<campaign_id>/export")
def export_campaign(campaign_id):
    conn = get_db()
    try:
        campaign = conn.execute(
            "SELECT id, name FROM campaigns WHERE id = ?", (campaign_id,)
        ).fetchone()
        if campaign is None:
            return jsonify(error="campaign not found"), 404

        characters = _count(conn, "campaign_characters", campaign_id)
        quests = _count(conn, "quests", campaign_id)
        npcs = _count(conn, "npcs", campaign_id)
        inventory_items = _count(conn, "campaign_inventory", campaign_id)
        sessions = _count(conn, "campaign_sessions", campaign_id)
    finally:
        conn.close()

    return jsonify(
        campaign_id=campaign["id"],
        name=campaign["name"],
        characters=characters,
        quests=quests,
        npcs=npcs,
        inventory_items=inventory_items,
        sessions=sessions,
        schema_version=SCHEMA_VERSION,
    )

File: results/dnd-rest-benchmark/lifecycle-runs/20260731T062307Z_claude_claude-sonnet-5_typescript-node/src/routes/audit.ts
// Deterministic audit and export summaries for campaign state. Read-only
// aggregation over existing campaign tables — no new persistent state.
import type { ServerResponse } from "node:http";
import { db, SCHEMA_VERSION } from "../db.js";
import { sendJson } from "../http.js";
import { hasCampaign } from "./campaigns.js";

function countRows(table: string, campaignId: string): number {
  const row = db.prepare(`SELECT COUNT(*) AS count FROM ${table} WHERE campaign_id = ?`).get(campaignId) as {
    count: number;
  };
  return row.count;
}

function countInventoryItems(campaignId: string): number {
  const row = db
    .prepare("SELECT COUNT(*) AS count FROM campaign_inventory WHERE campaign_id = ? AND countable = 1")
    .get(campaignId) as { count: number };
  return row.count;
}

function getCampaignName(campaignId: string): string {
  const row = db.prepare("SELECT name FROM campaigns WHERE id = ?").get(campaignId) as { name: string };
  return row.name;
}

export function handleGetCampaignAudit(res: ServerResponse, campaignId: string): void {
  if (!hasCampaign(campaignId)) {
    sendJson(res, 404, { error: "campaign not found" });
    return;
  }

  sendJson(res, 200, {
    campaign_id: campaignId,
    events: countRows("campaign_events", campaignId),
    quests: countRows("campaign_quests", campaignId),
    npcs: countRows("campaign_npcs", campaignId),
    sessions: countRows("campaign_sessions", campaignId),
  });
}

export function handleExportCampaign(res: ServerResponse, campaignId: string): void {
  if (!hasCampaign(campaignId)) {
    sendJson(res, 404, { error: "campaign not found" });
    return;
  }

  sendJson(res, 200, {
    campaign_id: campaignId,
    name: getCampaignName(campaignId),
    characters: countRows("campaign_characters", campaignId),
    quests: countRows("campaign_quests", campaignId),
    npcs: countRows("campaign_npcs", campaignId),
    inventory_items: countInventoryItems(campaignId),
    sessions: countRows("campaign_sessions", campaignId),
    schema_version: SCHEMA_VERSION,
  });
}
